{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4f969196",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: rdkit in /Users/samfields/Library/Python/3.9/lib/python/site-packages (2023.9.2)\n",
      "Requirement already satisfied: numpy in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from rdkit) (1.24.2)\n",
      "Requirement already satisfied: Pillow in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from rdkit) (10.1.0)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pyarrow in /Users/samfields/Library/Python/3.9/lib/python/site-packages (14.0.1)\n",
      "Requirement already satisfied: numpy>=1.16.6 in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from pyarrow) (1.24.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting umap\n",
      "  Downloading umap-0.1.1.tar.gz (3.2 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hBuilding wheels for collected packages: umap\n",
      "  Building wheel for umap (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for umap: filename=umap-0.1.1-py3-none-any.whl size=3564 sha256=ed36d8fb1da2ebf088c0add38a3a30b73cfa31f69a58423ad2de850c9d43f13c\n",
      "  Stored in directory: /Users/samfields/Library/Caches/pip/wheels/72/1e/42/a9322736ec046a637487005e6b84c94617d9ac0bdb5159eeb7\n",
      "Successfully built umap\n",
      "Installing collected packages: umap\n",
      "Successfully installed umap-0.1.1\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "# %apt install subversion\n",
    "# !svn checkout https://github.com/theislab/chemCPA.git\n",
    "%pip install rdkit\n",
    "%pip install pyarrow\n",
    "%pip install umap-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "666ac306",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting umap-learn\n",
      "  Downloading umap-learn-0.5.5.tar.gz (90 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m90.9/90.9 kB\u001b[0m \u001b[31m2.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: numpy>=1.17 in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from umap-learn) (1.24.2)\n",
      "Requirement already satisfied: scipy>=1.3.1 in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from umap-learn) (1.11.3)\n",
      "Requirement already satisfied: scikit-learn>=0.22 in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from umap-learn) (1.3.2)\n",
      "Collecting numba>=0.51.2\n",
      "  Downloading numba-0.58.1-cp39-cp39-macosx_11_0_arm64.whl (2.6 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.6/2.6 MB\u001b[0m \u001b[31m17.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting pynndescent>=0.5\n",
      "  Downloading pynndescent-0.5.11-py3-none-any.whl (55 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m55.8/55.8 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: tqdm in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from umap-learn) (4.66.1)\n",
      "Collecting llvmlite<0.42,>=0.41.0dev0\n",
      "  Downloading llvmlite-0.41.1-cp39-cp39-macosx_11_0_arm64.whl (28.8 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m28.8/28.8 MB\u001b[0m \u001b[31m48.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: joblib>=0.11 in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from pynndescent>=0.5->umap-learn) (1.3.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/samfields/Library/Python/3.9/lib/python/site-packages (from scikit-learn>=0.22->umap-learn) (3.2.0)\n",
      "Building wheels for collected packages: umap-learn\n",
      "  Building wheel for umap-learn (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for umap-learn: filename=umap_learn-0.5.5-py3-none-any.whl size=86854 sha256=468e11fb0d1a419bc68411e6353f32333992f771441f9dd50fe93808385f62ee\n",
      "  Stored in directory: /Users/samfields/Library/Caches/pip/wheels/0f/01/17/26951217a11fb724b7027c6dd5b620b6d368104f7e6d4171fc\n",
      "Successfully built umap-learn\n",
      "Installing collected packages: llvmlite, numba, pynndescent, umap-learn\n",
      "Successfully installed llvmlite-0.41.1 numba-0.58.1 pynndescent-0.5.11 umap-learn-0.5.5\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49m/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install umap-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "62679b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sklearn\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from rdkit.Chem import AllChem\n",
    "from rdkit import Chem\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "\n",
    "from sklearn.preprocessing import MinMaxScaler\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d805320b",
   "metadata": {},
   "source": [
    "# Template Python Notebook\n",
    "This notebook is meant to be an aggregate of all the code we wish to submit. No experimental code should be included in this notebook. Rather, you should duplicate this notebook when wisshing to develope new code, and after aggreement on the code and analysis the new code should then be copied into this notebook\n",
    "\n",
    "Competition Link: https://www.kaggle.com/competitions/open-problems-single-cell-perturbations/rules\n",
    "\n",
    "Single-Cell Perturbation Modeling Resource: https://www.sc-best-practices.org/conditions/perturbation_modeling.html "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a57f31d6",
   "metadata": {},
   "source": [
    "## Data Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "21c410a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH='open-problems-single-cell-perturbations'\n",
    "\n",
    "\n",
    "adata_obs_meta_path = PATH + '/adata_obs_meta.csv'\n",
    "adata_train_path = PATH + '/adata_train.parquet'\n",
    "de_train_path = PATH + '/de_train.parquet'\n",
    "id_map_path = PATH + '/id_map.csv'\n",
    "de_train_path = PATH + '/de_train.parquet'\n",
    "multiome_obs_meta_path = PATH + '/multiome_obs_meta.csv'\n",
    "multiome_train_path = PATH + '/multiome_train.parquet'\n",
    "multiome_var_meta_path = PATH + '/multiome_var_meta.csv'\n",
    "excluded_ids_path = PATH + '/adata_excluded_ids.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76c01605",
   "metadata": {},
   "source": [
    "## Data Preprocessing and Splitting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a52f77f",
   "metadata": {},
   "source": [
    "**`de_train.parquet`** - Aggregated differential expression data in dense array format.\n",
    "\n",
    "- `genes` A1BG, A1BG-AS1, …, ZZEF1 (numbering 18,211 in all) - Differential expression value (-log10(p-value) * sign(LFC)) for each gene. Here, LFC is the estimated log-fold change in expression between the treatment and control condition after shrinkage as calculated by Limma. Positive LFC means the gene goes up in the treatment condition relative to the control.\n",
    "- `cell_type` - The annotated cell type of each cell based on RNA expression.\n",
    "- `sm_name` - The primary name for the (parent) compound (in a standardized representation) as chosen by LINCS. This is provided to map the data in this experiment to the LINCS Connectivity Map data.\n",
    "- `sm_lincs_id` - The global LINCS ID (parent) compound (in a standardized representation). This is provided to map the data in this experiment to the LINCS Connectivity Map data.\n",
    "- `SMILES` - Simplified molecular-input line-entry system (SMILES) representations of the compounds used in the experiment. This is a 1D representation of molecular structure. These SMILES are provided by Cellarity based on the specific compounds ordered for this experiment.\n",
    "- `control` - Boolean indicating whether this instance was used as a control.\n",
    "\n",
    "**`adata_train.parquet `**- Unaggregated count and normalized data in COO sparse-array format. A supplement to de_train. In addition to the fields in de_train, this data also has:\n",
    "\n",
    "- `obs_id` - This is a unique identifier assigned to each cell in the raw dataset.\n",
    "- `gene` - Corresponds to the columns of de_train.\n",
    "- `count` - The raw molecular counts for the gene expression data measured in the experiment as output by 10x CellRanger.\n",
    "- `normalized_count` - These counts have been library size normalized and log(X+1) transformed.\n",
    "\n",
    "**`adata_obs_meta.csv`** - Observation metadata for adata_train.\n",
    "\n",
    "- `library_id` - A unique identifier for each library, which is a measurement made on pooled samples from each row of the plate. All cells from wells on the same row of the same plate will share a library_id.\n",
    "-` plate_name` - A unique ID for all samples from the same plate.\n",
    "well - The well location of the sample on each plate (this is standard across 96 well plate experiments). It is a concatenation of row and col.\n",
    "row - Which row on the plate the sample came from.\n",
    "- `col `- Which column on the plate the sample came from.\n",
    "donor_id - Identifies the donor source of the sample, one of three.\n",
    "- `cell_type` - The annotated cell type of each cell based on RNA expression. This matches the cell_type in the de_train.parquet.\n",
    "- `cell_id` - This is included for consistency with LINCS Connectivity Map metadata, which denotes a cell_id for each cell line.\n",
    "- `sm_name` - The primary name for the (parent) compound (in a standardized representation) as chosen by LINCS. This is provided to map the data in this experiment to the LINCS Connectivity Map data.\n",
    "- `sm_lincs_id` - The global LINCS ID (parent) compound (in a standardized representation). This is provided to map the data in this experiment to the LINCS Connectivity Map data.\n",
    "- `SMILES `- Simplified molecular-input line-entry system (SMILES) representations of the compounds used in the experiment. This is a 1D representation of molecular structure. These SMILES are provided by Cellarity based on the specific compounds ordered for this experiment.\n",
    "- `dose_uM` - Dose of the compound in on a micro-molar scale. This maps to the pert_idose field in LINCS.\n",
    "- timepoint_hr - Duration of treatment in hours. This maps to the pert_itime field in LINCS.\n",
    "- `control` - Whether this observation was used as a control, True or False.\n",
    "\n",
    "**`id_map.csv`** - Identifies the cell_type / sm_name pair to be predicted for the given id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bde26d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_train = pd.read_parquet(adata_train_path)\n",
    "multiome_train = pd.read_parquet(multiome_train_path)\n",
    "\n",
    "de_train = pd.read_parquet(de_train_path)\n",
    "\n",
    "adata_excluded_ids = pd.read_csv(excluded_ids_path)\n",
    "\n",
    "# Get gene list and convert to list\n",
    "genes_to_exclude = adata_excluded_ids['gene'].tolist()\n",
    "genes_to_exclude_set = set(genes_to_exclude)\n",
    "\n",
    "#Find the intersection with de_train columns and drop the columns from de_train\n",
    "genes_to_drop = genes_to_exclude_set & set(de_train.columns)\n",
    "de_train_filtered = de_train.drop(columns=list(genes_to_drop))\n",
    "\n",
    "adata_obs_meta =  pd.read_csv(adata_obs_meta_path)\n",
    "\n",
    "id_map = pd.read_csv(id_map_path)\n",
    "\n",
    "multiome_obs_meta = pd.read_csv(multiome_obs_meta_path)\n",
    "multiome_var_meta = pd.read_csv(multiome_var_meta_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f746ff91",
   "metadata": {},
   "source": [
    "## Single Cell Aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0cde50c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = de_train.iloc[:, 5:].columns.values.tolist() # Get all relevant genes\n",
    "multiome_train = multiome_train[multiome_train['location'].isin(genes)] # keep relevant genes in multiome\n",
    "\n",
    "multiome_train_ct = pd.merge(multiome_train, multiome_obs_meta, on='obs_id', how='left')\n",
    "multiome_train_ct_meta = pd.merge(multiome_train_ct, multiome_var_meta, left_on='location', \\\n",
    "                                    right_on='location', how='left')\n",
    "multiome_train_ct_meta_agg = multiome_train_ct_meta.groupby(['cell_type', 'location'])['normalized_count']\\\n",
    "    .agg(['mean', 'var']).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d9a61e02",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████| 6/6 [09:18<00:00, 93.05s/it]\n"
     ]
    }
   ],
   "source": [
    "transformed = []\n",
    "cell_types = multiome_train_ct_meta_agg.cell_type.unique()\n",
    "\n",
    "for cell_type in tqdm(cell_types):\n",
    "    row = {'cell_type': cell_type}\n",
    "    for gene in genes:\n",
    "        temp = multiome_train_ct_meta_agg[(multiome_train_ct_meta_agg['cell_type']==cell_type) & \\\n",
    "                                  (multiome_train_ct_meta_agg['location']==gene)]\n",
    "        mean = temp['mean'].values[0] if len(temp)>0 else 0\n",
    "        var = temp['var'].values[0] if len(temp)>0 else 0\n",
    "        row[f'{gene}_mean'] = mean\n",
    "        row[f'{gene}_var'] = var\n",
    "    transformed.append(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "acda7848",
   "metadata": {},
   "outputs": [],
   "source": [
    "multiome_train_ct_meta_agg_transformed = pd.DataFrame(transformed).fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b1585d",
   "metadata": {},
   "source": [
    "### Dimension Feature Reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "119f2105",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all columns except the first one\n",
    "columns_except_first = multiome_train_ct_meta_agg_transformed.columns[1:]\n",
    "\n",
    "# Store the first column separately\n",
    "first_column = multiome_train_ct_meta_agg_transformed[multiome_train_ct_meta_agg_transformed.columns[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63ced0b8",
   "metadata": {},
   "source": [
    "#### TSNE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ebc18c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tsne = TSNE(n_components=6, perplexity=5, method='exact', random_state=42)\n",
    "reduced = tsne.fit_transform(multiome_train_ct_meta_agg_transformed[columns_except_first])\n",
    "reduced = pd.DataFrame(reduced, columns=[f'SC Component {x}' for x in range(len(reduced[0]))])\n",
    "sc_pca = pd.concat([first_column, reduced], axis=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae8a6e26",
   "metadata": {},
   "source": [
    "#### PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "d4cc3a7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=6)  # You can choose the number of components you want to keep\n",
    "reduced = pca.fit_transform(multiome_train_ct_meta_agg_transformed[columns_except_first])\n",
    "reduced = pd.DataFrame(reduced, columns=[f'SC Component {x}' for x in range(len(reduced[0]))])\n",
    "sc_pca = pd.concat([first_column, reduced], axis=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ea1d880",
   "metadata": {},
   "source": [
    "#### UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "85afc139",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/samfields/Library/Python/3.9/lib/python/site-packages/umap/umap_.py:2433: UserWarning: n_neighbors is larger than the dataset size; truncating to X.shape[0] - 1\n",
      "  warn(\n"
     ]
    }
   ],
   "source": [
    "from umap import umap_ as UMAP\n",
    "# Create a UMAP object with the desired number of components\n",
    "umap_model = UMAP.UMAP(n_components=200, init='random')\n",
    "\n",
    "reduced = umap_model.fit_transform(multiome_train_ct_meta_agg_transformed[columns_except_first])\n",
    "reduced = pd.DataFrame(reduced, columns=[f'SC Component {x}' for x in range(len(reduced[0]))])\n",
    "sc_pca = pd.concat([first_column, reduced], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae8be7b8",
   "metadata": {},
   "source": [
    "### Compund Structure Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d5cf096a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mf_components = 200\n",
    "# Get SMILES embeddings\n",
    "de_train_filtered['Morgan_Fingerprints'] = de_train['SMILES'].apply(lambda x: AllChem.GetMorganFingerprintAsBitVect(Chem.MolFromSmiles(x), 2, nBits=1024))\n",
    "\n",
    "de_train_filtered['Morgan_Fingerprints'] = de_train_filtered['SMILES'].apply(lambda x: AllChem.GetMorganFingerprintAsBitVect(Chem.MolFromSmiles(x), 2, nBits=1024))\n",
    "# Perform PCA on SMILES embeddings\n",
    "pca = PCA(n_components=mf_components, random_state=42)\n",
    "features_pca = pca.fit_transform(list(de_train_filtered['Morgan_Fingerprints']))\n",
    "features_pca = pd.DataFrame(features_pca, columns=[f'MF PCA {x}' for x in range(len(features_pca[0]))])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76b0fad2",
   "metadata": {},
   "source": [
    "### Add Engineered Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "7837a497",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge MF components\n",
    "data = pd.merge(de_train_filtered, features_pca, left_index=True, right_index=True)\n",
    "\n",
    "# merge sc components\n",
    "data = pd.merge(data, sc_pca, on='cell_type', how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe5db4ce",
   "metadata": {},
   "source": [
    "### Normalize Engineered Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "b5cb915e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Initialize the MinMaxScaler\n",
    "# scaler = MinMaxScaler()\n",
    "\n",
    "# norm_cols = features_pca.columns.values.tolist()\n",
    "# norm_cols += sc_pca.columns.values.tolist()[1:]\n",
    "\n",
    "# # Normalize all columns in the DataFrame\n",
    "# data[norm_cols] = scaler.fit_transform(data[norm_cols])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20ee994",
   "metadata": {},
   "source": [
    "## Creating Training / Testing Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "00b45c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating test/train\n",
    "\n",
    "# Identify the cell types for training and testing\n",
    "training_cell_types = ['T cells (CD4+)', 'T cells (CD8+)', 'T cells (regulatory)', 'NK cells']\n",
    "testing_cell_types = ['Myeloid cells', 'B cells']\n",
    "\n",
    "# Filter data for training and testing\n",
    "train_data = data[data['cell_type'].isin(training_cell_types)]\n",
    "test_data = data[data['cell_type'].isin(testing_cell_types)]\n",
    "\n",
    "# Randomly select 50% of samples from Myeloid and B cells for testing\n",
    "test_samples = test_data.sample(frac=0.5, random_state=421)\n",
    "\n",
    "# The remaining samples are used for training\n",
    "train_samples = test_data.drop(test_samples.index)\n",
    "train_samples = pd.concat([train_data, train_samples])\n",
    "\n",
    "# declare features\n",
    "xfeatures = list(features_pca.columns) + list(reduced.columns) # + list(cell_type_ohe.columns)\n",
    "_yfeatures = ['cell_type', 'sm_name', 'sm_lincs_id', 'SMILES', 'control', 'Morgan_Fingerprints'] \\\n",
    "    + list(features_pca.columns) + list(reduced.columns) #+ list(cell_type_ohe.columns)\n",
    "\n",
    "# Want to predict gene expression from SMILES compound formation\n",
    "X_train = train_samples[xfeatures]\n",
    "y_train = train_samples.drop(columns=_yfeatures)\n",
    "\n",
    "X_test = test_samples[xfeatures]\n",
    "y_test = test_samples.drop(columns=_yfeatures)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b4a2ec5",
   "metadata": {},
   "source": [
    "## Export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "93c1aa71",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train.to_csv(\"Processed_Data/X_train.csv\", index=False)\n",
    "y_train.to_csv(\"Processed_Data/y_train.csv\", index=False)\n",
    "X_test.to_csv(\"Processed_Data/X_test.csv\", index=False)\n",
    "y_test.to_csv(\"Processed_Data/y_test.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "c95f91bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MF PCA 0</th>\n",
       "      <th>MF PCA 1</th>\n",
       "      <th>MF PCA 2</th>\n",
       "      <th>MF PCA 3</th>\n",
       "      <th>MF PCA 4</th>\n",
       "      <th>MF PCA 5</th>\n",
       "      <th>MF PCA 6</th>\n",
       "      <th>MF PCA 7</th>\n",
       "      <th>MF PCA 8</th>\n",
       "      <th>MF PCA 9</th>\n",
       "      <th>...</th>\n",
       "      <th>SC Component 190</th>\n",
       "      <th>SC Component 191</th>\n",
       "      <th>SC Component 192</th>\n",
       "      <th>SC Component 193</th>\n",
       "      <th>SC Component 194</th>\n",
       "      <th>SC Component 195</th>\n",
       "      <th>SC Component 196</th>\n",
       "      <th>SC Component 197</th>\n",
       "      <th>SC Component 198</th>\n",
       "      <th>SC Component 199</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.932676</td>\n",
       "      <td>-0.433296</td>\n",
       "      <td>0.143674</td>\n",
       "      <td>0.431457</td>\n",
       "      <td>-1.911878</td>\n",
       "      <td>0.413157</td>\n",
       "      <td>-0.500679</td>\n",
       "      <td>-0.754016</td>\n",
       "      <td>0.321070</td>\n",
       "      <td>0.096298</td>\n",
       "      <td>...</td>\n",
       "      <td>5.769604</td>\n",
       "      <td>5.749834</td>\n",
       "      <td>2.924859</td>\n",
       "      <td>6.165555</td>\n",
       "      <td>5.486909</td>\n",
       "      <td>4.192760</td>\n",
       "      <td>3.783000</td>\n",
       "      <td>3.587315</td>\n",
       "      <td>4.697591</td>\n",
       "      <td>5.569746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-1.225987</td>\n",
       "      <td>0.497097</td>\n",
       "      <td>-1.342844</td>\n",
       "      <td>-0.172098</td>\n",
       "      <td>2.219168</td>\n",
       "      <td>-0.597951</td>\n",
       "      <td>-0.216538</td>\n",
       "      <td>-0.328621</td>\n",
       "      <td>0.883519</td>\n",
       "      <td>0.383401</td>\n",
       "      <td>...</td>\n",
       "      <td>5.769604</td>\n",
       "      <td>5.749834</td>\n",
       "      <td>2.924859</td>\n",
       "      <td>6.165555</td>\n",
       "      <td>5.486909</td>\n",
       "      <td>4.192760</td>\n",
       "      <td>3.783000</td>\n",
       "      <td>3.587315</td>\n",
       "      <td>4.697591</td>\n",
       "      <td>5.569746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-0.669821</td>\n",
       "      <td>-0.761423</td>\n",
       "      <td>1.064209</td>\n",
       "      <td>-0.451881</td>\n",
       "      <td>-1.343732</td>\n",
       "      <td>0.269834</td>\n",
       "      <td>0.215291</td>\n",
       "      <td>3.235103</td>\n",
       "      <td>-1.560628</td>\n",
       "      <td>-1.341991</td>\n",
       "      <td>...</td>\n",
       "      <td>5.769604</td>\n",
       "      <td>5.749834</td>\n",
       "      <td>2.924859</td>\n",
       "      <td>6.165555</td>\n",
       "      <td>5.486909</td>\n",
       "      <td>4.192760</td>\n",
       "      <td>3.783000</td>\n",
       "      <td>3.587315</td>\n",
       "      <td>4.697591</td>\n",
       "      <td>5.569746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.610198</td>\n",
       "      <td>0.820041</td>\n",
       "      <td>0.085933</td>\n",
       "      <td>-1.925662</td>\n",
       "      <td>-0.385175</td>\n",
       "      <td>-0.356124</td>\n",
       "      <td>-1.813276</td>\n",
       "      <td>1.588276</td>\n",
       "      <td>-0.374456</td>\n",
       "      <td>0.785268</td>\n",
       "      <td>...</td>\n",
       "      <td>5.769604</td>\n",
       "      <td>5.749834</td>\n",
       "      <td>2.924859</td>\n",
       "      <td>6.165555</td>\n",
       "      <td>5.486909</td>\n",
       "      <td>4.192760</td>\n",
       "      <td>3.783000</td>\n",
       "      <td>3.587315</td>\n",
       "      <td>4.697591</td>\n",
       "      <td>5.569746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2.497142</td>\n",
       "      <td>1.749796</td>\n",
       "      <td>0.249717</td>\n",
       "      <td>-1.305727</td>\n",
       "      <td>-1.094619</td>\n",
       "      <td>-0.815843</td>\n",
       "      <td>-1.301174</td>\n",
       "      <td>-0.296460</td>\n",
       "      <td>0.586483</td>\n",
       "      <td>1.555864</td>\n",
       "      <td>...</td>\n",
       "      <td>5.769604</td>\n",
       "      <td>5.749834</td>\n",
       "      <td>2.924859</td>\n",
       "      <td>6.165555</td>\n",
       "      <td>5.486909</td>\n",
       "      <td>4.192760</td>\n",
       "      <td>3.783000</td>\n",
       "      <td>3.587315</td>\n",
       "      <td>4.697591</td>\n",
       "      <td>5.569746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>-0.616064</td>\n",
       "      <td>-0.344641</td>\n",
       "      <td>-0.419898</td>\n",
       "      <td>0.888072</td>\n",
       "      <td>0.741058</td>\n",
       "      <td>-0.711907</td>\n",
       "      <td>0.188368</td>\n",
       "      <td>-2.027478</td>\n",
       "      <td>0.554531</td>\n",
       "      <td>-0.644290</td>\n",
       "      <td>...</td>\n",
       "      <td>5.932283</td>\n",
       "      <td>5.776486</td>\n",
       "      <td>2.907589</td>\n",
       "      <td>6.170081</td>\n",
       "      <td>5.484545</td>\n",
       "      <td>4.201037</td>\n",
       "      <td>3.780656</td>\n",
       "      <td>3.605416</td>\n",
       "      <td>4.674687</td>\n",
       "      <td>5.630428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>-0.616064</td>\n",
       "      <td>-0.344641</td>\n",
       "      <td>-0.419898</td>\n",
       "      <td>0.888072</td>\n",
       "      <td>0.741058</td>\n",
       "      <td>-0.711907</td>\n",
       "      <td>0.188368</td>\n",
       "      <td>-2.027478</td>\n",
       "      <td>0.554531</td>\n",
       "      <td>-0.644290</td>\n",
       "      <td>...</td>\n",
       "      <td>6.003136</td>\n",
       "      <td>5.825334</td>\n",
       "      <td>2.877749</td>\n",
       "      <td>6.145342</td>\n",
       "      <td>5.468946</td>\n",
       "      <td>4.148154</td>\n",
       "      <td>3.791935</td>\n",
       "      <td>3.538408</td>\n",
       "      <td>4.742606</td>\n",
       "      <td>5.534284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>-0.082370</td>\n",
       "      <td>-0.992271</td>\n",
       "      <td>-0.945845</td>\n",
       "      <td>0.388957</td>\n",
       "      <td>0.416676</td>\n",
       "      <td>-0.201986</td>\n",
       "      <td>0.701888</td>\n",
       "      <td>0.764968</td>\n",
       "      <td>-0.018647</td>\n",
       "      <td>1.319008</td>\n",
       "      <td>...</td>\n",
       "      <td>5.932283</td>\n",
       "      <td>5.776486</td>\n",
       "      <td>2.907589</td>\n",
       "      <td>6.170081</td>\n",
       "      <td>5.484545</td>\n",
       "      <td>4.201037</td>\n",
       "      <td>3.780656</td>\n",
       "      <td>3.605416</td>\n",
       "      <td>4.674687</td>\n",
       "      <td>5.630428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>-0.082370</td>\n",
       "      <td>-0.992271</td>\n",
       "      <td>-0.945845</td>\n",
       "      <td>0.388957</td>\n",
       "      <td>0.416676</td>\n",
       "      <td>-0.201986</td>\n",
       "      <td>0.701888</td>\n",
       "      <td>0.764968</td>\n",
       "      <td>-0.018647</td>\n",
       "      <td>1.319008</td>\n",
       "      <td>...</td>\n",
       "      <td>6.003136</td>\n",
       "      <td>5.825334</td>\n",
       "      <td>2.877749</td>\n",
       "      <td>6.145342</td>\n",
       "      <td>5.468946</td>\n",
       "      <td>4.148154</td>\n",
       "      <td>3.791935</td>\n",
       "      <td>3.538408</td>\n",
       "      <td>4.742606</td>\n",
       "      <td>5.534284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>-0.064432</td>\n",
       "      <td>0.405555</td>\n",
       "      <td>0.920532</td>\n",
       "      <td>-1.067024</td>\n",
       "      <td>-0.407785</td>\n",
       "      <td>3.612149</td>\n",
       "      <td>0.728513</td>\n",
       "      <td>-0.673740</td>\n",
       "      <td>0.822107</td>\n",
       "      <td>0.574807</td>\n",
       "      <td>...</td>\n",
       "      <td>6.003136</td>\n",
       "      <td>5.825334</td>\n",
       "      <td>2.877749</td>\n",
       "      <td>6.145342</td>\n",
       "      <td>5.468946</td>\n",
       "      <td>4.148154</td>\n",
       "      <td>3.791935</td>\n",
       "      <td>3.538408</td>\n",
       "      <td>4.742606</td>\n",
       "      <td>5.534284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163 rows × 400 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     MF PCA 0  MF PCA 1  MF PCA 2  MF PCA 3  MF PCA 4  MF PCA 5  MF PCA 6  \\\n",
       "0   -0.932676 -0.433296  0.143674  0.431457 -1.911878  0.413157 -0.500679   \n",
       "4   -1.225987  0.497097 -1.342844 -0.172098  2.219168 -0.597951 -0.216538   \n",
       "10  -0.669821 -0.761423  1.064209 -0.451881 -1.343732  0.269834  0.215291   \n",
       "14   1.610198  0.820041  0.085933 -1.925662 -0.385175 -0.356124 -1.813276   \n",
       "18   2.497142  1.749796  0.249717 -1.305727 -1.094619 -0.815843 -1.301174   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "517 -0.616064 -0.344641 -0.419898  0.888072  0.741058 -0.711907  0.188368   \n",
       "518 -0.616064 -0.344641 -0.419898  0.888072  0.741058 -0.711907  0.188368   \n",
       "535 -0.082370 -0.992271 -0.945845  0.388957  0.416676 -0.201986  0.701888   \n",
       "536 -0.082370 -0.992271 -0.945845  0.388957  0.416676 -0.201986  0.701888   \n",
       "557 -0.064432  0.405555  0.920532 -1.067024 -0.407785  3.612149  0.728513   \n",
       "\n",
       "     MF PCA 7  MF PCA 8  MF PCA 9  ...  SC Component 190  SC Component 191  \\\n",
       "0   -0.754016  0.321070  0.096298  ...          5.769604          5.749834   \n",
       "4   -0.328621  0.883519  0.383401  ...          5.769604          5.749834   \n",
       "10   3.235103 -1.560628 -1.341991  ...          5.769604          5.749834   \n",
       "14   1.588276 -0.374456  0.785268  ...          5.769604          5.749834   \n",
       "18  -0.296460  0.586483  1.555864  ...          5.769604          5.749834   \n",
       "..        ...       ...       ...  ...               ...               ...   \n",
       "517 -2.027478  0.554531 -0.644290  ...          5.932283          5.776486   \n",
       "518 -2.027478  0.554531 -0.644290  ...          6.003136          5.825334   \n",
       "535  0.764968 -0.018647  1.319008  ...          5.932283          5.776486   \n",
       "536  0.764968 -0.018647  1.319008  ...          6.003136          5.825334   \n",
       "557 -0.673740  0.822107  0.574807  ...          6.003136          5.825334   \n",
       "\n",
       "     SC Component 192  SC Component 193  SC Component 194  SC Component 195  \\\n",
       "0            2.924859          6.165555          5.486909          4.192760   \n",
       "4            2.924859          6.165555          5.486909          4.192760   \n",
       "10           2.924859          6.165555          5.486909          4.192760   \n",
       "14           2.924859          6.165555          5.486909          4.192760   \n",
       "18           2.924859          6.165555          5.486909          4.192760   \n",
       "..                ...               ...               ...               ...   \n",
       "517          2.907589          6.170081          5.484545          4.201037   \n",
       "518          2.877749          6.145342          5.468946          4.148154   \n",
       "535          2.907589          6.170081          5.484545          4.201037   \n",
       "536          2.877749          6.145342          5.468946          4.148154   \n",
       "557          2.877749          6.145342          5.468946          4.148154   \n",
       "\n",
       "     SC Component 196  SC Component 197  SC Component 198  SC Component 199  \n",
       "0            3.783000          3.587315          4.697591          5.569746  \n",
       "4            3.783000          3.587315          4.697591          5.569746  \n",
       "10           3.783000          3.587315          4.697591          5.569746  \n",
       "14           3.783000          3.587315          4.697591          5.569746  \n",
       "18           3.783000          3.587315          4.697591          5.569746  \n",
       "..                ...               ...               ...               ...  \n",
       "517          3.780656          3.605416          4.674687          5.630428  \n",
       "518          3.791935          3.538408          4.742606          5.534284  \n",
       "535          3.780656          3.605416          4.674687          5.630428  \n",
       "536          3.791935          3.538408          4.742606          5.534284  \n",
       "557          3.791935          3.538408          4.742606          5.534284  \n",
       "\n",
       "[163 rows x 400 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acbaea71",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
